pyccel/codegen/printing/ccode.py

```diff
@@ -1344,9 +1344,12 @@ def _print_Interface(self, expr):
     def _print_FunctionDef(self, expr):
         if expr.is_inline:
             return ''
-
-        if len(expr.results) > 1:
+        results = expr.results
+        if len(results) > 1:
             self._additional_args.append(expr.results)
+        if isinstance(results[0], Variable) and results[0].allocatable:
+            results[0].allocatable = False
+            results[0].is_pointer = True
```

Comment:
Why do you need a pointer? Why only the 1st object that is returned? What happens if we do:
```
def f():
    import numpy as np
    a = np.ones(5)
    b = np.ones(3)
    return a,b
```

Reply:
you're right, we have no need to use a pointer.
This code should be at the semantic stage, to prevent creating the allocate nodes for the assigned objects to functions that return an allocatable object (for both C and Fortran). that's why we should disable `allocatable` for the assigned objects by disabling it for the returned objects of functions, The problem is this breaks Fortran, causing a wrong data-typing for the return object.

Here is the current behavior of allocatable:
python:
```python
from numpy import ones

def create_array():
	z = ones(5)
	return z

def return_array():
	a = create_array()
	b = ones(5)
	return a

if __name__ == '__main__':
	c = return_array()
	print(c)
```
C:
module:
```C
#include "poo.h"
#include <stdlib.h>
#include "ndarrays.h"


/*........................................*/
t_ndarray create_array(void)
{
    t_ndarray z = {.shape = NULL};
    z = array_create(1, (int64_t[]){5}, nd_double);
    array_fill((double)1.0, z);
    return z;
}
/*........................................*/
/*........................................*/
t_ndarray return_array(void)
{
    t_ndarray b = {.shape = NULL};
    t_ndarray a = {.shape = NULL};
    a = array_create(1, (int64_t[]){5}, nd_double);
    a = create_array();
    b = array_create(1, (int64_t[]){5}, nd_double);
    array_fill((double)1.0, b);
    free_array(b);
    return a;
}
/*........................................*/
```
the program:
```C
#include "poo.h"
#include <stdlib.h>
#include "ndarrays.h"
#include <stdio.h>
#include <stdint.h>
int main()
{
    t_ndarray c = {.shape = NULL};
    int64_t i;
    c = array_create(1, (int64_t[]){5}, nd_double);
    c = return_array();
    printf("%s", "[");
    for (i = 0; i < 4; i += 1)
    {
        printf("%.12lf ", GET_ELEMENT(c, nd_double, (int64_t)i));
    }
    printf("%.12lf]\n", GET_ELEMENT(c, nd_double, (int64_t)4));
    free_array(c);
    return 0;
}
```
the same problem in Fortran (creating a `deallocate` node for an object that is already assigned to a function that returns an allocatable object)

Here is what happens if we modify `allocatable` attribute in a `Variable` that will be returned in a function at the assigning stage (in this case, we have `a = create_array()` and `c = return_array()`)

the C code will be perfect, which means deallocate nodes won't be created for `a` (at `return_array`) and (`c` at the main function), and `a` and `c` will be declared properly.

for Fortran it's a bit problematic,
here is the generated code:
the module:
```Fortran
module poo


  use, intrinsic :: ISO_C_Binding, only : f64 => C_DOUBLE , i64 => &
        C_INT64_T
  implicit none

  contains

  !........................................
  function create_array() result(z)

    implicit none

    real(f64) :: z(0:4_i64)

    allocate(z(0:4_i64))
    z = 1.0_f64
    return

  end function create_array
  !........................................

  !........................................
  function return_array() result(a)

    implicit none

    real(f64) :: a(0:4_i64)
    real(f64), allocatable :: b(:)

    a = create_array()
    allocate(b(0:4_i64))
    b = 1.0_f64
    if (allocated(b)) then
      deallocate(b)
    end if
    return

  end function return_array
  !........................................

end module poo
```

the program:
```Fortran
program prog_prog_poo

  use poo

  use, intrinsic :: ISO_C_Binding, only : f64 => C_DOUBLE , i64 => &
        C_INT64_T
  implicit none

  real(f64) :: c(0:4_i64)

  c = return_array()
  print *, c
  if (allocated(c)) then
    deallocate(c)
  end if

end program prog_prog_poo
```

`a` and `c` are not declared properly, they are missing just the allocatable attribute

Reply:
i think we should modify `allocatable` after the object gets the right declaration format

Reply:
> to prevent creating the allocate nodes for the assigned objects to functions that return an allocatable object

I presume you mean `deallocate` nodes?

I am a little confused. It seems you are setting `allocatable = False` (means size is known or variable points to another variable) in order to avoid the creation of a `deallocate` node. Doesn't [this line](https://github.com/pyccel/pyccel/blob/return_arrays/pyccel/parser/semantic.py#L2841) prevent the creation of the `deallocate` node?

I am not sure what has happened in the program to get an array that is not allocatable but has a `deallocate` node

Reply:
🤦‍♂️ im so sorry for not mentioning "memory leaks", i was playing around it 😅.
in the examples above of the C code, these two lines
```C
    a = array_create(1, (int64_t[]){5}, nd_double);
    a = create_array();
```
are causing memory leaks, the same thing for
```C
    c = array_create(1, (int64_t[]){5}, nd_double);
    c = return_array();
```

Reply:
So we need to not allocate what is already allocated inside a function

Reply:
Ah you are right. Isn't this happening **because** you have set `allocatable=False`? As the code knows that `allocatable` won't be called it is initialising the data straight away. It is the equivalent of `real(f64) :: c(0:4_i64)` in fortran where the data has also already been initialised to a known size

Reply:
for C it's working perfectly, but fortran seems to need extra work, I think we should keep the memory leaks fix at the semantic since it's for both C and Fortran, and I hope there is a way to fix that Fortran problem at the printer,
what do you think ?

Reply:
If there are memory leaks then it is not working perfectly. I think we should have `allocatable = True`. I think having `allocatable = False` is causing the memory leaks as well as the fortran problems. The `deallocate` at the end of the function should be handled where it is created not by creating a new kind of array with:
- allocatable=False
- pointer=False
- stack_array=False

as I can't tell how this array is supposed to be initialised